hysr/hysr_control.py

```diff
@@ -384,3 +389,285 @@ def natural_reset(
+      string. If the latest, it should evaluate to a dict. Instances will be created
+      by importing the class from the module, and instantiating the class passing the
+      provided args and kargs. The classes should be subclasses of hysr.PressureRobot
+      and hysr.MainSim.
+    - 'algorithm_time_step' and "mujoco_time_step": should be float (in seconds)
+    Optional are the keys 'extra_balls' and 'imports':
+    - 'extra_balls': If provided, the value for 'extra_balls'
+      should be a dict, with each entry providing the configuration required to instantiate
+      hysr.ExtraBalls (or a subclass of it). The key of each entry is arbitrary and will
+      not be used.
+    - 'imports': list of packages (as strings) to import before constructing the instances. May be useful
+      if the args and kwargs are string to evaluate.
+    """
+
+    try:
+        t = toml.loads(toml_string)
+    except toml.TomlDecodeError as tde:
+        raise ValueError(f"hysr control factory: invalid toml content: {tde}")
+
+    def _read_factory_class(
+        toml_content: typing.Dict[str, typing.Any], key: str
+    ) -> hysr_types.FactoryClass:
+        if key not in toml_content.keys():
+            raise ValueError(
+                f"hysr control factory: failed to find required key {key} in toml configuration"
+            )
+        d = toml_content[key]
+        required_keys = ("class",)
+        optional_keys = ("args", "kwargs")
+        for rk in required_keys:
+            if rk not in d.keys():
+                raise ValueError(
+                    f"hysr control factory: failed to find required key {key}/{rk} in toml configuration"
+                )
+        for k in d.keys():
+            if k not in required_keys + optional_keys:
+                raise ValueError(
+                    f"hysr control factory: the toml configuration has an unexpected key: {key}/{k}"
+                )
+        class_ = str(d["class"])
+        if "args" in d.keys():
+            if isinstance(d["args"], str):
+                try:
+                    args = eval(d["args"])
+                except NameError as ne:
+                    raise ValueError(
+                        f"hysr_control_factory: failed to evaluate args {d['args']}: {ne} "
+                    )
+            else:
+                args = d["args"]
+            if not isinstance(args, collections.abc.Sequence):
+                raise ValueError(
+                    f"hysr control factory: the toml configuration has an unexpected value for: {key}/args (should be a sequence)"
+                )
+        else:
+            args = []
+        if "kwargs" in d.keys():
+            if isinstance(d["kwargs"], str):
+                kwargs = eval(d["kwargs"])
+            else:
+                kwargs = d["kwargs"]
+        else:
+            kwargs = {}
+
+        return (class_, args, kwargs)
+
+    # importing requested imports, if any
```

Comment:
Shouldn't the imports be done before `eval`-ing the args to be useful?

Reply:
this is the case, the eval is done in the function, which is declared before the import, but called after them

Reply:
Ah, sorry, I missed that.